include/nbl/asset/IRayTracingPipeline.h

```diff
@@ -0,0 +1,171 @@
+    };
+};
+
+template<typename PipelineLayoutType, typename ShaderType>
+class IRayTracingPipeline : public IPipeline<PipelineLayoutType>, public IRayTracingPipelineBase
+{
+  public:
+
+    using SGeneralShaderGroupContainer = core::smart_refctd_dynamic_array<SGeneralShaderGroup>;
+    using SHitShaderGroupContainer = core::smart_refctd_dynamic_array<SHitShaderGroup>;
+
+    struct SCreationParams : IPipeline<PipelineLayoutType>::SCreationParams
+    {
+      protected:
+        using SpecInfo = ShaderType::SSpecInfo;
+        template<typename ExtraLambda>
+        inline bool impl_valid(ExtraLambda&& extra) const
+        {
+          if (!IPipeline<PipelineLayoutType>::SCreationParams::layout)
+            return false;
+
+          core::bitflag<ICPUShader::E_SHADER_STAGE> stagePresence = {};
+          for (const auto info : shaders)
+            if (info.shader)
+            {
+              if (!extra(info))
+                return false;
+              const auto stage = info.shader->getStage();
+              if (stage > ICPUShader::E_SHADER_STAGE::ESS_CALLABLE || stage < ICPUShader::E_SHADER_STAGE::ESS_RAYGEN)
+                return false;
+              if (stage == ICPUShader::E_SHADER_STAGE::ESS_RAYGEN && stagePresence.hasFlags(hlsl::ESS_RAYGEN))
+                return false;
+              stagePresence |= stage;
+            }
+
+          auto getShaderStage = [this](size_t index) -> ICPUShader::E_SHADER_STAGE
+            {
+              return shaders[index].shader->getStage();
+            };
+
+          if (shaderGroups.raygenGroup.shaderIndex >= shaders.size())
+            return false;
+          if (getShaderStage(shaderGroups.raygenGroup.shaderIndex) != ICPUShader::E_SHADER_STAGE::ESS_RAYGEN)
+            return false;
+
+          auto isValidShaderIndex = [this, getShaderStage](size_t index, ICPUShader::E_SHADER_STAGE expectedStage) -> bool
+            {
+              if (index == SShaderGroupsParams::ShaderUnused)
+                return true;
+              if (index >= shaders.size())
+                return false;
+              if (getShaderStage(index) != expectedStage)
+                return false;
+              return true;
+            };
+
+          for (const auto& shaderGroup : shaderGroups.hitGroups)
+          {
+            if (!isValidShaderIndex(shaderGroup.anyHitShaderIndex, ICPUShader::E_SHADER_STAGE::ESS_ANY_HIT))
+              return false;
+
+            if (!isValidShaderIndex(shaderGroup.closestHitShaderIndex, ICPUShader::E_SHADER_STAGE::ESS_CLOSEST_HIT))
+              return false;
+
+            if (!isValidShaderIndex(shaderGroup.intersectionShaderIndex, ICPUShader::E_SHADER_STAGE::ESS_INTERSECTION))
+              return false;
+          }
+
+          for (const auto& shaderGroup : shaderGroups.missGroups)
+          {
+            if (!isValidShaderIndex(shaderGroup.shaderIndex, ICPUShader::E_SHADER_STAGE::ESS_MISS))
+              return false;
+          }
```

Comment:
I need to read vulkan spec and find out what's the point of multiple miss groups

Reply:
> The ray payload for a shader invoked by a ray tracing command must be the same structure for all shader stages using the payload for that ray.

I guess its to allow multiple different payload structs